pinc/NonactivatedUser.inc

```diff
@@ -0,0 +1,188 @@
+    );
+
+    public function __construct($username=NULL)
+    {
+        if($username)
+        {
+            $this->load("username", $username);
+        }
+    }
+
+    // The __set() and __get() methods allow access to user fields without
+    // creating accessors for them all individually.
+    // See the PHP docs for "magic methods".
+    public function __set($name, $value)
+    {
+        if(in_array($name, $this->unsettable_fields))
+        {
+            throw new DomainException(sprintf(
+                _("%s is an unsettable field"), $name)
+            );
+        }
+
+        if(isset($this->$name) && in_array($name, $this->immutable_fields))
+        {
+            throw new DomainException(sprintf(
+                _("%s is an immutable field"), $name)
+            );
+        }
+
+        $this->table_row[$name] = $value;
+    }
+
+    public function __get($name)
+    {
+        return $this->table_row[$name];
+    }
+
+    public function __isset($name)
+    {
+        return isset($this->table_row[$name]);
+    }
+
+    public function load($field, $value, $strict=TRUE)
+    {
+        if(!in_array($field, $this->integer_fields))
+        {
+            $escaped_value = sprintf("'%s'", mysqli_real_escape_string(DPDatabase::get_connection(), $value));
+        }
+        else
+        {
+            $escaped_value = $value;
+        }
+
+        $sql = "
+            SELECT *
+            FROM non_activated_users
+            WHERE $field=$escaped_value
+        ";
+
+        $result = mysqli_query(DPDatabase::get_connection(), $sql);
+        if(!$result)
+        {
+            throw new UnexpectedValueException(mysqli_error(DPDatabase::get_connection()));
+        }
+        elseif(mysqli_num_rows($result) == 0)
+        {
+            throw new NonexistentNonactivatedUserException(sprintf(
+                _('No non_activated_user found with %1$s = %2$s'),
+                    $field, $value)
+            );
+        }
+        elseif(mysqli_num_rows($result) > 1)
+        {
+            throw new NonuniqueNonactivatedUserException(sprintf(
+                _('Multiple non_activated_users found with %1$s = %2$s'),
+                    $field, $value)
+            );
+        }
+        $this->table_row = mysqli_fetch_assoc($result);
+
+        # MySQL does case-insensitive matches for strings, as well as
+        # ignores trailing whitespace. If $strict is TRUE, and the field
+        # is the username, require that the requested $username matches the
+        # existing username in case and whitespace by doing a PHP string
+        # compare.
+        if($strict and $field == 'username' and $this->username != $value)
+        {
+            throw new NonexistentNonactivatedUserException(sprintf(
+                _('No non_activated_user found with %1$s = %2$s'),
+                    $field, $value)
+            );
+        }
+
+        mysqli_free_result($result);
+    }
+
+    public function save()
```

Comment:
I think you are technically correct, however the odds of two non-bots attempting to register the same username at the same time is ridiculously small.

Reply:
Yeah, probably. We do occasionally receive duplicated requests (from non-bots), which is a more likely scenario for hitting this race condition -- but since the requests would be identical, you'd get a successful + correct registration regardless?

Anyhow, I think the "ON DUPLICATE KEY" deserves a comment in the code: someone could reasonably look at that and think "wait, if the key already exists in the table, isn't that a bug?" And once you've explained why it's there, you could maybe mention that it also allows a ridiculously unlikely race condition.

Reply:
I've updated the save unit test to detect this, confirmed it was broken, updated the code, and confirmed it is fixed.

Reply:
> The primary key for that row is the code-generated id (userid + 32-character hex string)

Hm? In the `non_activated_users` table, the primary key is the `username` column, not the `id` column. Or am I looking at the wrong thing?

Reply:
You are correct, I don't know what I was looking at.

Reply:
(So then what did the unit test detect?)

Reply:
The unit test detected that we were always creating a new ID upon every save and that we no longer do that now.

Reply:
Updated function comment with mention about race condition since the primary key is the username.